File: cpp/benchmarks/raw_frame_queue_benchmark.cpp
#include "llab/raw_frame_queue.hpp"

#include <algorithm>
#include <array>
#include <atomic>
#include <chrono>
#include <cmath>
#include <cstdint>
#include <iomanip>
#include <iostream>
#include <pthread.h>
#include <sched.h>
#include <thread>
#include <vector>

#if defined(__x86_64__) || defined(_M_X64)
#include <immintrin.h>
#endif

namespace {
constexpr std::uint64_t events = 1'000'000;
constexpr std::size_t samples = 20;
constexpr std::size_t payload_capacity = 4096;
constexpr int producer_cpu = 0;
constexpr int consumer_cpu = 2;

double rank(std::vector<double> values, const double percentile) {
    std::sort(values.begin(), values.end());
    return values[static_cast<std::size_t>(std::ceil(values.size() * percentile)) - 1];
}

void spin_pause() {
#if defined(__x86_64__) || defined(_M_X64)
    _mm_pause();
#else
    std::atomic_signal_fence(std::memory_order_seq_cst);
#endif
}

bool pin_current_thread(const int cpu) {
    cpu_set_t set;
    CPU_ZERO(&set);
    CPU_SET(cpu, &set);
    return pthread_setaffinity_np(pthread_self(), sizeof(set), &set) == 0;
}
}  // namespace

int main() {
    std::vector<double> durations;
    durations.reserve(samples);
    for (std::size_t sample = 0; sample < samples; ++sample) {
        (void)sample;
        llab::RawFrameSpscQueue queue(1024, payload_capacity);
        std::atomic<bool> start = false;
        std::atomic<unsigned int> ready = 0;
        std::atomic<std::uint64_t> consumed = 0;
        std::thread producer([&] {
            if (!pin_current_thread(producer_cpu)) std::terminate();
            ready.fetch_add(1, std::memory_order_release);
            const std::array<std::uint8_t, 64> small{};
            const std::array<std::uint8_t, 2048> large{};
            while (!start.load(std::memory_order_acquire)) spin_pause();
            for (std::uint64_t index = 0; index < events; ++index) {
                const auto payload = index % 16 == 0 ? std::span<const std::uint8_t>(large) : std::span<const std::uint8_t>(small);
                const llab::RawFrameHeader header{index, index, 0, 1, llab::FrameDirection::Inbound, llab::FrameKind::Text};
                while (queue.try_push(header, payload) != llab::RawFramePushResult::Queued) spin_pause();
            }
        });
        std::thread consumer([&] {
            if (!pin_current_thread(consumer_cpu)) std::terminate();
            ready.fetch_add(1, std::memory_order_release);
            llab::RawFrameRecord output{0, 0, 0, 0, llab::FrameDirection::Inbound, llab::FrameKind::Text, {}};
            output.payload.reserve(payload_capacity);
            while (!start.load(std::memory_order_acquire)) spin_pause();
            for (std::uint64_t expected = 0; expected < events; ++expected) {
                while (!queue.try_pop_into(output)) spin_pause();
                if (output.capture_index != expected) std::terminate();
            }
            consumed.store(events, std::memory_order_release);
        });
        while (ready.load(std::memory_order_acquire) != 2) spin_pause();
        const auto begun = std::chrono::steady_clock::now();
        start.store(true, std::memory_order_release);
        producer.join();
        consumer.join();
        if (consumed.load(std::memory_order_acquire) != events) std::terminate();
        durations.push_back(std::chrono::duration<double, std::nano>(std::chrono::steady_clock::now() - begun).count() / events);
    }
    std::cout << std::fixed << std::setprecision(3)
              << "samples=" << samples << " events_per_sample=" << events
              << " producer_cpu=" << producer_cpu << " consumer_cpu=" << consumer_cpu
              << " payload=15x64B+1x2048B p50_ns_per_message=" << rank(durations, 0.50)
              << " p99_ns_per_message=" << rank(durations, 0.99) << '\n';
}

File: rust/src/bin/raw_frame_queue_benchmark.rs
use low_latency_lab_benchmarks::raw_frame_queue::{RawFrameHeader, RawFrameSpscQueue};
use low_latency_lab_benchmarks::raw_frame_record::RawFrameRecord;
use std::sync::Arc;
use std::sync::atomic::{AtomicBool, AtomicU64, Ordering};
use std::thread;
use std::time::Instant;

const EVENTS: u64 = 1_000_000;
const SAMPLES: usize = 20;
const PAYLOAD_CAPACITY: usize = 4096;

fn rank(values: &mut [f64], percentile: f64) -> f64 {
    values.sort_by(f64::total_cmp);
    values[((values.len() as f64 * percentile).ceil() as usize - 1).min(values.len() - 1)]
}

fn main() {
    let mut samples = Vec::with_capacity(SAMPLES);
    for _ in 0..SAMPLES {
        let queue = Arc::new(RawFrameSpscQueue::new(1024, PAYLOAD_CAPACITY));
        let start = Arc::new(AtomicBool::new(false));
        let consumed = Arc::new(AtomicU64::new(0));
        let producer_queue = Arc::clone(&queue);
        let producer_start = Arc::clone(&start);
        let producer = thread::spawn(move || {
            let small = [0xA5_u8; 64];
            let large = [0x5A_u8; 2048];
            while !producer_start.load(Ordering::Acquire) {
                std::hint::spin_loop();
            }
            for capture_index in 0..EVENTS {
                let payload = if capture_index % 16 == 0 { &large[..] } else { &small[..] };
                let header = RawFrameHeader {
                    capture_index,
                    monotonic_ns: capture_index,
                    utc_ns: 0,
                    connection_id: 1,
                    direction: 0,
                    kind: 0,
                };
                while producer_queue.try_push(header, payload).is_err() {
                    std::hint::spin_loop();
                }
            }
        });
        let consumer_queue = Arc::clone(&queue);
        let consumer_start = Arc::clone(&start);
        let consumer_count = Arc::clone(&consumed);
        let consumer = thread::spawn(move || {
            let mut out = RawFrameRecord {
                capture_index: 0,
                monotonic_ns: 0,
                utc_ns: 0,
                connection_id: 0,
                direction: 0,
                kind: 0,
                payload: Vec::with_capacity(PAYLOAD_CAPACITY),
            };
            while !consumer_start.load(Ordering::Acquire) {
                std::hint::spin_loop();
            }
            for expected in 0..EVENTS {
                while !consumer_queue.try_pop_into(&mut out) {
                    std::hint::spin_loop();
                }
                assert_eq!(out.capture_index, expected);
            }
            consumer_count.store(EVENTS, Ordering::Release);
        });
        let started = Instant::now();
        start.store(true, Ordering::Release);
        producer.join().unwrap();
        consumer.join().unwrap();
        assert_eq!(consumed.load(Ordering::Acquire), EVENTS);
        samples.push(started.elapsed().as_nanos() as f64 / EVENTS as f64);
    }
    let mut p50 = samples.clone();
    let mut p99 = samples.clone();
    println!(
        "samples={} events_per_sample={} payload=15x64B+1x2048B p50_ns_per_message={:.3} p99_ns_per_message={:.3}",
        SAMPLES,
        EVENTS,
        rank(&mut p50, 0.50),
        rank(&mut p99, 0.99)
    );
}
